dipy/reconst/odffp.py

```diff
@@ -0,0 +1,1038 @@
+'''
+Created on Feb 18, 2021
+
+@author: patrykfi
+'''
+
+import os, gzip
+import h5py, hdf5storage
+import warnings
+
+import numpy as np
+import nibabel as nib
+
+import matplotlib.pyplot as plt
+import mpl_toolkits.mplot3d.art3d as art3d
+import scipy.stats
+
+from dipy.core.sphere import Sphere
+from dipy.core.dsi_sphere import dsiSphere8Fold
+from dipy.core.geometry import sphere2cart
+
+from dipy.direction import peak_directions
+
+from dipy.reconst.gqi import GeneralizedQSamplingModel
+from dipy.reconst.odf import OdfFit
+from dipy.reconst.shm import sf_to_sh, sh_to_sf
+
+from scipy.io import loadmat, savemat
+
+
+DEFAULT_RECON_EDGE = 1.2
+DEFAULT_DICT_EDGE = 1.2
+
+DEFAULT_FIT_PENALTY = 0.00001
+MAX_FIT_PENALTY = 0.1
+
+
+def plot_odf(odf, filename='odf.png', tessellation=dsiSphere8Fold()):
+    
+    fig = plt.figure(figsize=(30, 10))
+    
+    elev = [90, 90, 0]
+    azim = [-90, 0, -90]
+    
+    for i in range(3):
+        ax = fig.add_subplot(1, 3, i+1, projection="3d")
+        norm = plt.Normalize(odf.min(), odf.max())
+        v = (tessellation.vertices.T * norm(odf)).T
+        f = tessellation.faces
+    
+        colors = np.mean(abs(v[f]),axis=1)
+        norm = plt.Normalize(colors.min(), colors.max())
+        colors = norm(colors)
+    
+        pc = art3d.Poly3DCollection(v[f], facecolors=colors, edgecolor=None)
+    
+        ax.add_collection(pc)
+        ax.set_xlim3d([-1, 1])
+        ax.set_ylim3d([-1, 1])
+        ax.set_zlim3d([-1, 1])
+     
+        ax.view_init(elev[i], azim[i])
+    
+    fig.savefig(filename, bbox_inches='tight')
+    
+    plt.close(fig)    
+
+
+# Base abstract class
+class DiffusionDataGenerator(object):
```

Comment:
All the generator functions/class might move to the simulation module ``dipy.sims``.